### PR TITLE
Allow an attribute with default boolean value to be set correctly

### PR DESCRIPTION
**Pull request by da-ar**

### CODE REVIEW

File: lib/puppet/resource_api.rb [hunk 1]
```diff
@@ -178,9 +178,14 @@ def to_resource
             isnamevar
           end
 
+          type = Puppet::ResourceApi.parse_puppet_type(name, options[:type])
+
           # read-only values do not need type checking, but can have default values
-          if options[:behaviour] != :read_only
-            if options.key? :default
+          if options[:behaviour] != :read_only && options.key?(:default)
+            if type.is_a? Puppet::Pops::Types::PBooleanType
```
Comment (DavidS) [inline lib/puppet/resource_api.rb]: What about compound types like `Optional[Boolean]`, `Array[Boolean]` and `Variant[String, Boolean]`?

File: lib/puppet/resource_api.rb [hunk 2]
```diff
@@ -521,6 +524,15 @@ def self.mungify(type, value, error_msg_prefix)
       # the correct types, e.g. "1" (a string) to 1 (an integer)
       cleaned_value, error_msg = try_mungify(type, value, error_msg_prefix)
       raise Puppet::ResourceError, error_msg if error_msg
+    elsif type.is_a? Puppet::Pops::Types::PBooleanType
```
Comment (DavidS) [inline lib/puppet/resource_api.rb]: See compound types above.

File: spec/puppet/resource_api_spec.rb [hunk 3]
```diff
@@ -1034,6 +1052,74 @@ def set(_context, _changes); end
...
+      end
+
+      it { expect { described_class.register_type(definition) }.not_to raise_error }
+      it { expect(instance.flush).to eq(final_hash) }
+    end
+
+    context 'when the default value is false' do
+      let(:type) { Puppet::Type.type(:default_bool) }
+      let(:default_value) { false }
+      let(:instance) { type.new(name: 'foo', ensure: 'present') }
```
Comment (DavidS) [inline spec/puppet/resource_api_spec.rb]: lots of duplication here that can be pulled out to the parent context?

File: spec/puppet/resource_api_spec.rb [hunk 4]
```diff
@@ -1344,6 +1430,7 @@ def set(_context, changes)
         context 'when flushing' do
           before(:each) do
             Puppet.debug = true
+            instance.my_provider.set(nil, nil) # reset the current_state
```
Comment (DavidS) [inline spec/puppet/resource_api_spec.rb]: Huh, how are the changes above leaking into this test case?

File: lib/puppet/resource_api.rb (hunk 1 above)
Comment (DavidS) [reply lib/puppet/resource_api.rb]: Would it be enough to apply the workaround if and only if the default value is actually `false`?

File: spec/puppet/resource_api_spec.rb (hunk 4 above)
Comment (da-ar) [reply spec/puppet/resource_api_spec.rb]: believe random seed order was changed when the new test went in and exposed the bug

Comment (DavidS) [reply spec/puppet/resource_api_spec.rb]: ah, ok! then please put this in a separate commit with a short note about that, to inform future-us of what went down here.

**Comment by DavidS** [comment]
Added some cleanups to the tests. After this has passed CI, we can merge and release?

**Pull request merged**